darts/datasets/dataset_loaders.py

```diff
@@ -0,0 +1,147 @@
+from dataclasses import dataclass
+import os
+from pathlib import Path
+import hashlib
+from abc import ABC, abstractmethod
+
+import pandas as pd
+import requests
+
+from ..timeseries import TimeSeries
+
+
+@dataclass
+class DatasetLoaderMetadata:
+    # name of the dataset file, including extension
+    name: str
+    # uri of the dataset, expects a publicly available file
+    uri: str
+    # md5 hash of the file to be downloaded
+    hash: str
+    # used to parse the dataset file
+    header_time: str
+    # used to convert the string date to pd.Datetime
+    # https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior
+    format_time: str = None
+
+
+class DatasetLoadingException(BaseException):
+    pass
+
+
+class DatasetLoader(ABC):
+    """
+    Class that downloads a dataset and caches it locally.
+    Assumes that the file can be downloaded (i.e. publicly available via an URI)
+    """
+    _DEFAULT_DIRECTORY = Path(os.path.join(Path.home(), Path('.darts/datasets/')))
+
+    def __init__(self, metadata: DatasetLoaderMetadata, root_path: Path = None):
+        self._metadata: DatasetLoaderMetadata = metadata
+        if root_path is None:
+            self._root_path: Path = DatasetLoader._DEFAULT_DIRECTORY
+        else:
+            self._root_path: Path = root_path
+
+    def load(self) -> TimeSeries:
+        """
+        Load the dataset in memory, as a TimeSeries.
```

Comment:
It doesn't have to be supported right now, but it would be nice at some point to also support datasets that consist in a `Sequence[TimeSeries]` (e.g., M4 competition dataset).